src/Framework/Sdk/SdkResultItem.cs

```diff
@@ -27,42 +33,30 @@ public SdkResultItem()
         public SdkResultItem(string itemSpec, Dictionary<string, string> metadata)
         {
             ItemSpec = itemSpec;
-            Metadata = metadata;
+            Metadata = metadata ?? new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
         }
 
         public override bool Equals(object obj)
         {
             if (obj is SdkResultItem item &&
                    ItemSpec == item.ItemSpec &&
-                   Metadata?.Count == item.Metadata?.Count)
+                   Metadata?.Count == item.Metadata.Count)
```

Comment:
```suggestion
                   Metadata?.Count == item.Metadata?.Count)
```

You'll presumably need this too before removing the `?? new Dictionary...` part.

Reply:
@Forgind 

Let me see... So do you think Metadata may be passed with null value on the 2nd constructor that has Metadata as parameter?
If this is true, then I have to update the SdkResultItem unit test of this to ShouldNotBe like this (using diff):

``` diff
        [Fact]
        public void SdkResultItem_Equal_CtorParam_MetadataNull()
        {
            var sdkResultItem1 = new SdkResultItem("anyspec", new Dictionary<string, string>());
            var sdkResultItem2 = new SdkResultItem("anyspec", null);

-            // Should bt the same, because passed null metadata will have value of new Dictionnary<String,String> like sdkResultItem1
-            sdkResultItem1.ShouldBe(sdkResultItem2);
+           // Should not be the same, because passing metadata = null is allowed and the Metadata property value allows null.
+           sdkResultItem1.ShouldNotBe(sdkResultItem2);
        }
```

Please confirm. Thanks in advance!

Reply:
Yes, correct. We want to allow Metadata to be null.

Reply:
ok, I have updated the code (but I haven't pushed yet) so now it looks like this:

```
        public override bool Equals(object obj)
        {
            if (obj is SdkResultItem item &&
                   ItemSpec == item.ItemSpec &&
                   Metadata?.Count == item.Metadata?.Count)
            {
                return Metadata.All(m => item.Metadata.TryGetValue(m.Key, out var itemValue) && itemValue == m.Value);
            }
            return false;
        }
```
But now this line: 
```
return Metadata.All(m => item.Metadata.TryGetValue(m.Key, out var itemValue) && itemValue == m.Value);
```
cause this warning: `Dereference of a possibly null reference.` especially on `Item.Metadata.TryGetValue(...` which I think that warning will be treated as Error in the CI.
If I change to `item.Metadata?.TryGetValue(...` then I will get error of can't convert bool? to bool again.

Please confirm. I'm a bit confused now, because now I'm back to previous compile errors I mentioned before (both deference and the bool? error)

Reply:
Ok, so we want Metadata to be able to be null. The compiler (with nullable enabled) looks for anything that can possibly be null and ensures that you verify that it isn't null before you try to dereference it because otherwise you could get an exception at runtime.

If Metadata can be null, then calling `TryGetValue` on item.Metadata is calling a function from something that might be null—it's dereferencing a null value.

When you add the `?`, that checks if it's null and only calls the function afterwards if it isn't. If it is, it skips the function call and replace it with null.

Because we are in a nullable-enabled environment, the compiler is still looking for things that can possibly be null and annotating them as such. TryGetValue normally returns a normal bool, i.e., not null. Because we had item.Metadata?.TryGetValue, it will either return a bool from TryGetValue _or_ skip the TryGetValue and return null because Metadata is null. Returning something that is either null or a bool is returning a bool? or nullable bool.

The && operator cannot take in null because it isn't exactly true but it also isn't exactly false. We need a way to convert a bool? into a bool without possibly throwing an error if it is null.

Normally, it's pointless to say, for instance, `if (<bool value> == True)` because the bool value is already either true or false. However, with a bool?, it can be true, false, or null. Checking explicitly if it == True (or == False) is a check that it is _not_ null _and_ that it is the value you're checking. For them to be equal, we need all the metadata to be the same, which means it cannot be null if our metadata is not null, and it must have the value, so TryGetValue has to return true. Checking item.Metadata?.TryGetValue(m.Key, out var itemValue) == True covers both of those cases at the same time.

If you want to make this easier to understand, you can also explicitly check if either is null at the top, then cast them to non-null Dictionaries and compare them as such.

Does that make more sense?

Reply:
@Forgind 

Many thanks! Makes sense now, and my confusion is gone.

based on this:
> If you want to make this easier to understand, you can also explicitly check if either is null at the top, then cast them to non-null Dictionaries and compare them as such.

I have updated that code to be like this checking of `item.Metadata  == null` is within the if clauses) since this also remove the dereference warning (imho the logic feels cleaner cmiiw)

```
        public override bool Equals(object obj)
        {
            if (obj is SdkResultItem item &&
                   ItemSpec == item.ItemSpec &&
                   item.Metadata != null &&
                   Metadata?.Count == item.Metadata.Count)
            {
                return Metadata.All(m => item.Metadata.TryGetValue(m.Key, out var itemValue) && itemValue == m.Value);
            }
            return false;
        }
```
If you are ok, I will commit that with the updated unit test to use `ShouldNotBe`.

Again, many thanks for the explanation and patience!

Reply:
I think I've committed my proposal now, including the updated unit test in commit https://github.com/dotnet/msbuild/pull/6839/commits/8593b36bec85853bf7234864ded7cee6c92d6c86

So we could quickly examine the change I proposed. Also I need to check the CI so to ensure I have no compile errors.

If you have additional feedback, I'll update them again.

Thanks in advance!

Reply:
It looks like you're still asserting that Metadata is not null by declaring it with:
`public Dictionary<string, string> Metadata { get;}`
instead of
`public Dictionary<string, string>? Metadata { get;}`

If you add that in and don't give it a starting value if it's null, it should break again at the top of equals because you never verify that this.Metadata is not null. (Also, tiny nit: you can technically override the `==` method, so we prefer to use `is null` and `is not null`.)

Reply:
Nice catch! thanks!

I've updated the code.

Reply:
@Forgind 

Done updating the code and applying to use `is not null`. CI is green.
Please review.

Hopefully this meet the feedback 🙂 
Thanks!

Reply:
@Forgind 

Done. Thanks

Reply:
Looks good. If this passes CI, I'm happy 🙂

Reply:
CI is all green now 🙂

Reply:
@Forgind 
Now that the CI is all green, is there anything else I should check?

Reply:
just to nudge/remind, 
anything else for me to check? or do I have to wait for other PR to be reviewed and merged before this PR?
sorry to always asking you folks. 🙏 
it's ok if this PR may wait for other PRs or if you're busy. 🙂 
because my team need this fix, so hopefully next week I could use new msbuild that has this fix.

okay, I'm going to step back and check this PR tomorrow. :wave:
many thanks for your guidance and patience, folks!

@Forgind  @rainersigwald @BenVillalobos

Reply:
> my team need this fix

Can you elaborate on how you're seeing this at all? From the original bug it looked like a code-inspection discovered problem that was purely theoretical.

Reply:
@rainersigwald 
I need this to compare `SdkResultItem` reliably, because ,my team and I have some scenarios that use comparison of this SdkResultItem. 
Sorry, I could not disclose the further detail of the scenario. 🙏 

I'll use my own msbuild repo while waiting for this to be reviewed further, then. No need to rush, folks.
Thanks again! 👍

Reply:
I'd wait for rainersigwald to sign off too; if he does, we can merge this.